motorola-potter: init

telent opened the pull request:
Based closely on addison, but with a  different kernel upstream

* boots into 'hello' example
* framebuffer works
* adb works
* touchscreen works
* wlan0 works (use the addison process to get modem firmware), tested with wpa_supplicant/dhcp/ping 

nixpkgs hash 900115a4f7fdd9189e7803ca781a65be663f2c89 (23 May)

samueldr commented on devices/motorola-potter/default.nix:
```diff
@@ -0,0 +1,71 @@
+
+  hardware.firmware = [
+    (config.mobile.device.firmware.override {
+      modem = ../../vendor/modem;
```
Probably won't work for other people :)

telent replied:
no, you're right. will delete before merge

samueldr commented on devices/motorola-potter/default.nix:
```diff
@@ -0,0 +1,71 @@
+    };
+  };
+
+  # Using `xz` allows us to fit the kernel + dt + initrd in 16MiB
+  #   Kernel: ~8.8M
+  #   DT:      1.2M
+  # We're left with ~6MB for the compressed initrd.
+  mobile.boot.stage-1.compression = lib.mkDefault "xz";
```
Since you were talking about the boot.img being too big, I assume this is not the numbers for your system.

samueldr commented on devices/motorola-potter/kernel/config.armv7l:
```diff
@@ -0,0 +1,4851 @@
+#
```
Delete this file, unless you have validated with armv7l...

(I probably should for addison...)

samueldr commented on devices/motorola-potter/kernel/default.nix:
```diff
@@ -0,0 +1,42 @@
+, ...
+}:
+
+#
+# Note:
+# This kernel build is special, it supports both armv7l and aarch64.
+# This is because motorola ships an armv7l userspace from stock ROM.
+#
+# in local.nix:
+#  mobile.system.system = lib.mkForce "armv7l-linux";
+#
```
Similarly, unless you have validated it works just as well for armv7l, remove this. (And `
stdenv.hostPlatform.parsed.cpu.name` use...)

telent replied:
gone

samueldr commented on devices/motorola-potter/partitions_32GB.gdisk:
```diff
@@ -0,0 +1,76 @@
+GPT fdisk (gdisk) version 1.0.4
```
Dumped from your device?

telent replied:
was leftover from addison, but now replaced with a fresh one

samueldr reviewed (commented):
Misc things, I expect some were just "TODO" not done yet on your end :).

telent commented:
Just for completeness sake, some numbers for bot.img size:
```
-r--r--r-- 1 root root 18151424 Jan  1  1970 /nix/store/wzyhviakn0sbc82dzq3jqd1fswlw00nz-mobile-nixos_motorola-potter_boot.img
-r--r--r-- 1 root root 18118656 Jan  1  1970 /nix/store/30rncw27lybk0jw6r5ksc5zx21l25sp8-mobile-nixos_motorola-potter_boot.img -CONFIG_AUDIT
-r--r--r-- 1 root root 18468864 Jan  1  1970 /nix/store/6zfy0s6ljip3vnvpcwa7qg0f9jasx7p5-mobile-nixos_motorola-potter_boot.img busybox musl
-r--r--r-- 1 root root 17631232 Jan  1  1970 /nix/store/j5rj0k8s4bdv3vjhf7r06k697f39n281-mobile-nixos_motorola-potter_boot.img busybox minimal (!musl)
-r--r--r-- 1 root root 17160192 Jan  1  1970 /nix/store/vbpzxxh7wnwllycg0nybsgmvbdxgqgc7-mobile-nixos_motorola-potter_boot.img actually fixing the DTBs (!minimal)
-r--r--r-- 1 root root 16510976 Jan  1  1970 /nix/store/r2lkfisj1p97i9029qcq9gqf5f3hgmc9-mobile-nixos_motorola-potter_boot.img sysemdMinimal for udev in initrd
```

telent commented:
Touchscreen works, boot.img < 16MB, wifi works, previous comments addressed - is there anything else I need to look at to make this mergable?

telent commented on modules/initrd.nix:
```diff
@@ -1,12 +1,12 @@
 { config, pkgs, lib, utils, ... }:
 
 let
+  udev = pkgs.systemdMinimal;
```
Worth drawing your attention to this, as it's core code not port code. I don't *imagine* that using systemdMinimal for udev in the initrd should break any other device, but that might be just my lack of imagination

samueldr replied:
I don't think so either since we're cherry-picking the actual udev description files.

samueldr commented on overlay/mobile-nixos/kernel/builder.nix:
```diff
@@ -203,7 +206,14 @@ stdenv.mkDerivation (inputArgs // {
+    # (e.g. qcom/foo.dtb shoudl end up in arch/arm/boot/dts/qcom/foo.dtb).
+    # The options are to ensure the files/dirs are created writable so that
+    # sed can edit them in postPatch
+    tar -C ${outOfTreeDtbs} -cf - . | tar -C arch/arm/boot/dts/ --mode=a+w --no-overwrite-dir -xvvpf -
```
Aren't those `outOfTreeDts`es?

We probably want to copy the structure at the root. AArch64 dts would be at `arch/arm64/boot/dts/qcom`... and then there's non-qualcomm device trees.

If instead the out of tree DTSes were rooted at `arch/arm/boot/dts/qcom/*.dts` that would make it work in a more universal manner.

telent replied:
> AArch64 dts would be at `arch/arm64/boot/dts/qcom`

Hmm. This *is* an aarch64 device, but the kernel build expects them in `arm` not `arm64`. Maybe that's unique to this kernel lineage and not a generic thing.

I think you're right to be unsure,  it feels like there are too many variables involved to make this generic - given, as you say, that a patch is needed anyway

samueldr replied:
Exactly, it's because it's an AArch64 with an ARM soul :).

Though *some* actually-AArch64 devices *do* put DTS files in `arch/arm/boot/dts`, which makes this **even more confusing**!!

telent replied:
64c1ae4 fixes this

samueldr reviewed (commented):
Some initial notes

samueldr commented:
I'm not sure about the out of tree DTS thing. I don't dislike the concept, but I think in your situation it would be better served with a proper `.patch` that requires no change in the tooling. Since *anyway* the way it ends up working in your case is through a kernel patch.

Does that make sense?

telent commented on modules/quirks/qualcomm/touchscreen-powerup.nix:
```diff
@@ -0,0 +1,34 @@
+      wantedBy = [ "multi-user.target" ];
+      script = ''
+        # would be better if we could detect this path somehow
+        ts_device=/sys/devices/soc/78b7000.i2c/i2c-3/3-0020
```
Fair points, and I'll check but I'm pretty sure it will need to be a Task. For reference, https://wiki.postmarketos.org/wiki/Troubleshooting#Touch_screen_device_present_but_not_outputting_any_touch_events  lists two other Motorola devices that are said to need this fix

samueldr replied:
I *think* the `fb-notify` hack actually does this through another "cleaner" route, but I'm not 100% sure.

telent replied:
fixed in kernel, no longer needed

telent commented on devices/motorola-potter/default.nix:
```diff
@@ -0,0 +1,75 @@
+    "androidboot.bootdevice=7824900.sdhci"
+    "lpm_levels.sleep_disabled=1"
+    "vmalloc=350M"
+    "buildvariant=userdebug"
```
none required, all deleted

telent commented on devices/motorola-potter/default.nix:
```diff
@@ -0,0 +1,75 @@
+  mobile.system.type = "android";
+
+  mobile.quirks.qualcomm = {
+    fb-notify.enable = true; # copied from addison, don't know if needed
```
no longer needed, fixed in kernel

samueldr commented on devices/motorola-potter/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+  mobile.device.firmware = pkgs.callPackage ./firmware {};
+
+  # WiFi needs firmware from the device. Refer to README.adoc in this
+  # directory for instructions on how to get it, then uncomment this
+  # and change the path appropriately
+
+  # hardware.firmware = [
+  #   (config.mobile.device.firmware.override {
+  #     modem = ../../vendor/modem;
+  #   })
+  # ];
+
```
The end-user shouldn't need to edit this file.

This should be in the `README.adoc` file. Though to be fair generic firmware hints should also exist independent from all devices.


```suggestion
```

telent replied:
For clarity: the user should be putting these settings in their configuration.nix instead of here?

samueldr replied:
Yeah, in their system configuration (e.g. `/etc/nixos/configuration.nix`) or if they're building outside of the system, they can add it to the system configuration, (e.g. in `local.nix`).

samueldr commented on devices/motorola-potter/default.nix:
```diff
@@ -0,0 +1,61 @@
+
+  mobile.hardware = {
+    soc = "qualcomm-msm8953";
+    ram = 1024 * 3;
```
```suggestion
    ram = 1024 * 2;
```

According to [this page](https://www.gsmarena.com/motorola_moto_g5_plus-8453.php), the "Moto G5 Plus" exists in variants with 2, 3 and 4 GiB of RAM.

This documents the smallest available RAM size, but is currently totally unused.

I *assume* it might be used soon for the installer, to provide hints at build time to the user that the "copytoram" installer variants may not work.

Otherwise setting a lower amount than your actual device has *has no impact*.

samueldr commented on devices/motorola-potter/README.adoc:
```diff
@@ -0,0 +1,97 @@
+boot image to fit in the allotted 16MB partition we have merged the
+duplicates.
+
+hwrev  board-id     (mis)features
+
+P0A  <0x44 0x80a0>   U D
+P1A  <0x44 0x81a0>   U D
+P1B  <0x44 0x81b0>   D
+P2A  <0x44 0x82a0>   D
+     <0x45 0x82a0>   D
+P2A2 <0x45 0x82a2> 
+P2B  <0x45 0x82b0> 
+P3A  <0x46 0x83a0> 
+P3B  <0x46 0x83b0> 
+
```
This either needs to be a table, or a pre-formatted text area.

samueldr commented on devices/motorola-potter/README.adoc:
```diff
@@ -0,0 +1,97 @@
+=== Hardware Revisions
+
+This device has been through several hardware revisions: the kernel
+describes P0A, P1A, P1B,P2A, P2A2, P2B, and P3A but that's not an
```
I personally would use pre-formatted text for the device variants.

```suggestion
describes `P0A`, `P1A`, `P1B`, `P2A`, `P2A2`, `P2B`, and `P3A` but that's not an
```

samueldr commented on devices/motorola-potter/README.adoc:
```diff
@@ -0,0 +1,97 @@
+- msm8953-potter-p2a2-plus.dtb (contains no fix)
+
+
+==== What about P3B?
```
```suggestion
==== What about `P3B`?
```

samueldr commented on devices/motorola-potter/README.adoc:
```diff
@@ -0,0 +1,97 @@
+==== What about P3B?
+
+The vendor kernel does not include a device tree source with
+appropriate board-id for hwrev P3B, but empirically (by consulting
```
```suggestion
appropriate board-id for hwrev `P3B`, but empirically (by consulting
```

samueldr commented on devices/motorola-potter/README.adoc:
```diff
@@ -0,0 +1,97 @@
+The vendor kernel does not include a device tree source with
+appropriate board-id for hwrev P3B, but empirically (by consulting
+`/proc/device-tree/board-id` on a device running Android) it looks like
+the bootloader chooses the P3A DTB instead. As far as I understand it
```
```suggestion
the bootloader chooses the `P3A` DTB instead. As far as I understand it
```

samueldr commented on devices/motorola-potter/README.adoc:
```diff
@@ -0,0 +1,97 @@
+for choosing a device tree - https://android.googlesource.com/kernel/msm.git/+/android-msm-bullhead-3.10-n-preview-1/Documentation/devicetree/bindings/arm/msm/board-id.txt?autodive=0%2F%2F%2F%2F[according to
+sources]
+the platform subtype is the lowest 8 bits of the second element of the
+board-id, and clearly 0xa0 != 0xb0, but it seems to do it anyway.
```
```suggestion
`board-id`, and clearly `0xa0 != 0xb0`, but it seems to do it anyway.
```

samueldr commented on devices/motorola-potter/README.adoc:
```diff
@@ -0,0 +1,97 @@
+the platform subtype is the lowest 8 bits of the second element of the
+board-id, and clearly 0xa0 != 0xb0, but it seems to do it anyway.
+
+If you have a device with a hwrev *after* P3B, I'd love to hear from
```
```suggestion
If you have a device with a hwrev *after* `P3B`, I'd love to hear from
```

samueldr commented on overlay/mobile-nixos/kernel/builder.nix:
```diff
@@ -203,7 +203,7 @@ stdenv.mkDerivation (inputArgs // {
     fi
   ''
-    + maybeString prePatch
+  + maybeString prePatch
```
Since this is not touching the kernel builder anymore, drop this change.

(Re-formatting to be in-line with how everything else is done would be welcome if this file was changed otherwise.)

samueldr reviewed (changes requested):
Some minor changes.

Looking good on the whole, lots of nits :).

I haven't taken a look at the commits history yet, but when you're fixing up these changes, you can take the time to cleanup what isn't needed. But as we discussed, please keep some of your steps and stumbles visible, as it has values in showing how a port can be made.

telent commented:
Hi Samuel - think I've addressed all open comments, please take another look

I've done some (light, minor) cleanups of the timeline, based on the criterion:
- technical wrong turns and discoveries (e.g the weird touchscreen stuff) I left intact
- simple "process" wrong turns (out of tree DTS, or adding cruft in motorola-potter/default.nix) I've fixed

samueldr commented:
Note to self: I need to pick an arbitrary device and build with this PR applied to verify the changes to the package selected for udev in stage-1.

samueldr commented:
♥ Thank you!
